subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java

```diff
@@ -69,9 +79,11 @@ public FileSystemChangeWaiter createChangeWaiter(PendingChangesListener listener
         private final Collection<FileWatcherEvent> eventsBeforeListening = new ArrayList<FileWatcherEvent>();
         private final Lock eventDeliveryLock = new ReentrantLock();
 
-        private ChangeWaiter(FileWatcherFactory fileWatcherFactory, final PendingChangesListener pendingChangesListener, long quietPeriodMillis, BuildCancellationToken cancellationToken) {
+        private ChangeWaiter(FileWatcherFactory fileWatcherFactory, final PendingChangesListener pendingChangesListener, long quietPeriodMillis, boolean triggeredBuild, BuildCancellationToken cancellationToken, BuildGateToken buildGateToken) {
```

Comment:
I don't like pushing `buildGateToken.waitForOpen()` into here. The simplest improvement would be to wrap this `ChangeWaiter` in a `BuildGatedChangeWaiter`.

But even better if we split the `FileSystemChangeWaiter` into 2 things: a basic `ContinuousBuildMonitor` API that has `wait(Runnable notifier)` and another that has the various filesystem specific methods. This change might be out of scope for the current story.